// @effect/platform/dist/esm/HttpMultiplex.js
import * as internal from "./internal/httpMultiplex.js";
/**
 * @since 1.0.0
 * @category type ids
 */
export const TypeId = internal.TypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const empty = internal.empty;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = internal.make;
/**
 * @since 1.0.0
 * @category combinators
 */
export const add = internal.add;
/**
 * @since 1.0.0
 * @category combinators
 */
export const headerExact = internal.headerExact;
/**
 * @since 1.0.0
 * @category combinators
 */
export const headerRegex = internal.headerRegex;
/**
 * @since 1.0.0
 * @category combinators
 */
export const headerStartsWith = internal.headerStartsWith;
/**
 * @since 1.0.0
 * @category combinators
 */
export const headerEndsWith = internal.headerEndsWith;
/**
 * @since 1.0.0
 * @category combinators
 */
export const hostExact = internal.hostExact;
/**
 * @since 1.0.0
 * @category combinators
 */
export const hostRegex = internal.hostRegex;
/**
 * @since 1.0.0
 * @category combinators
 */
export const hostStartsWith = internal.hostStartsWith;
/**
 * @since 1.0.0
 * @category combinators
 */
export const hostEndsWith = internal.hostEndsWith;
