// @effect/platform/dist/esm/HttpClientResponse.js
import * as internal from "./internal/httpClientResponse.js";
export {
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyJson,
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyJsonScoped,
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyUrlParams,
/**
 * @since 1.0.0
 * @category schema
 */
schemaBodyUrlParamsScoped,
/**
 * @since 1.0.0
 * @category schema
 */
schemaHeaders,
/**
 * @since 1.0.0
 * @category schema
 */
schemaHeadersScoped } from "./HttpIncomingMessage.js";
/**
 * @since 1.0.0
 * @category type ids
 */
export const TypeId = internal.TypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const fromWeb = internal.fromWeb;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaJson = internal.schemaJson;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaNoBody = internal.schemaNoBody;
/**
 * @since 1.0.0
 * @category accessors
 */
export const arrayBuffer = internal.arrayBuffer;
/**
 * @since 1.0.0
 * @category accessors
 */
export const formData = internal.formData;
/**
 * @since 1.0.0
 * @category accessors
 */
export const json = internal.json;
const void_ = internal.void_;
export {
/**
 * @since 1.0.0
 * @category accessors
 */
void_ as void };
/**
 * @since 1.0.0
 * @category accessors
 */
export const stream = internal.stream;
/**
 * @since 1.0.0
 * @category accessors
 */
export const text = internal.text;
/**
 * @since 1.0.0
 * @category accessors
 */
export const urlParamsBody = internal.urlParamsBody;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaJsonScoped = internal.schemaJsonScoped;
/**
 * @since 1.0.0
 * @category schema
 */
export const schemaNoBodyScoped = internal.schemaNoBodyScoped;
/**
 * @since 1.0.0
 * @category pattern matching
 */
export const matchStatus = internal.matchStatus;
/**
 * @since 1.0.0
 * @category pattern matching
 */
export const matchStatusScoped = internal.matchStatusScoped;
