// @effect/platform/dist/esm/Worker.js
import * as internal from "./internal/worker.js";
/**
 * @since 1.0.0
 * @category type ids
 */
export const PlatformWorkerTypeId = internal.PlatformWorkerTypeId;
/**
 * @since 1.0.0
 */
export const makePlatform = internal.makePlatform;
/**
 * @since 1.0.0
 * @category tags
 */
export const PlatformWorker = internal.PlatformWorker;
/**
 * @since 1.0.0
 * @category tags
 */
export const Spawner = internal.Spawner;
/**
 * @since 1.0.0
 * @category type ids
 */
export const WorkerManagerTypeId = internal.WorkerManagerTypeId;
/**
 * @since 1.0.0
 * @category tags
 */
export const WorkerManager = internal.WorkerManager;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeManager = internal.makeManager;
/**
 * @since 1.0.0
 * @category layers
 */
export const layerManager = internal.layerManager;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makePool = internal.makePool;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makePoolLayer = internal.makePoolLayer;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeSerialized = internal.makeSerialized;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makePoolSerialized = internal.makePoolSerialized;
/**
 * @since 1.0.0
 * @category layers
 */
export const makePoolSerializedLayer = internal.makePoolSerializedLayer;
/**
 * @since 1.0.0
 * @category layers
 */
export const layerSpawner = internal.layerSpawner;
