// effect/dist/esm/internal/schema/filters.js
/** @internal */
export const GreaterThanSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThan");
/** @internal */
export const GreaterThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanOrEqualTo");
/** @internal */
export const LessThanSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThan");
/** @internal */
export const LessThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanOrEqualTo");
/** @internal */
export const IntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Int");
/** @internal */
export const BetweenSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Between");
/** @internal */
export const GreaterThanBigintSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanBigint");
/** @internal */
export const GreaterThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanOrEqualToBigint");
/** @internal */
export const LessThanBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanBigint");
/** @internal */
export const LessThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanOrEqualToBigint");
/** @internal */
export const BetweenBigintSchemaId = /*#__PURE__*/Symbol.for("effect/schema/BetweenBigint");
/** @internal */
export const MinLengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MinLength");
/** @internal */
export const MaxLengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MaxLength");
/** @internal */
export const LengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Length");
/** @internal */
export const MinItemsSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MinItems");
/** @internal */
export const MaxItemsSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MaxItems");
/** @internal */
export const ItemsCountSchemaId = /*#__PURE__*/Symbol.for("effect/schema/ItemsCount");
