// effect/dist/cjs/internal/schema/filters.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MinLengthSchemaId = exports.MinItemsSchemaId = exports.MaxLengthSchemaId = exports.MaxItemsSchemaId = exports.LessThanSchemaId = exports.LessThanOrEqualToSchemaId = exports.LessThanOrEqualToBigIntSchemaId = exports.LessThanBigIntSchemaId = exports.LengthSchemaId = exports.ItemsCountSchemaId = exports.IntSchemaId = exports.GreaterThanSchemaId = exports.GreaterThanOrEqualToSchemaId = exports.GreaterThanOrEqualToBigIntSchemaId = exports.GreaterThanBigintSchemaId = exports.BetweenSchemaId = exports.BetweenBigintSchemaId = void 0;
/** @internal */
const GreaterThanSchemaId = exports.GreaterThanSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThan");
/** @internal */
const GreaterThanOrEqualToSchemaId = exports.GreaterThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanOrEqualTo");
/** @internal */
const LessThanSchemaId = exports.LessThanSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThan");
/** @internal */
const LessThanOrEqualToSchemaId = exports.LessThanOrEqualToSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanOrEqualTo");
/** @internal */
const IntSchemaId = exports.IntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Int");
/** @internal */
const BetweenSchemaId = exports.BetweenSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Between");
/** @internal */
const GreaterThanBigintSchemaId = exports.GreaterThanBigintSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanBigint");
/** @internal */
const GreaterThanOrEqualToBigIntSchemaId = exports.GreaterThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/GreaterThanOrEqualToBigint");
/** @internal */
const LessThanBigIntSchemaId = exports.LessThanBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanBigint");
/** @internal */
const LessThanOrEqualToBigIntSchemaId = exports.LessThanOrEqualToBigIntSchemaId = /*#__PURE__*/Symbol.for("effect/schema/LessThanOrEqualToBigint");
/** @internal */
const BetweenBigintSchemaId = exports.BetweenBigintSchemaId = /*#__PURE__*/Symbol.for("effect/schema/BetweenBigint");
/** @internal */
const MinLengthSchemaId = exports.MinLengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MinLength");
/** @internal */
const MaxLengthSchemaId = exports.MaxLengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MaxLength");
/** @internal */
const LengthSchemaId = exports.LengthSchemaId = /*#__PURE__*/Symbol.for("effect/schema/Length");
/** @internal */
const MinItemsSchemaId = exports.MinItemsSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MinItems");
/** @internal */
const MaxItemsSchemaId = exports.MaxItemsSchemaId = /*#__PURE__*/Symbol.for("effect/schema/MaxItems");
/** @internal */
const ItemsCountSchemaId = exports.ItemsCountSchemaId = /*#__PURE__*/Symbol.for("effect/schema/ItemsCount");
