// @effect/platform/dist/cjs/HttpServerError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isServerError = exports.exitResponse = exports.clientAbortFiberId = exports.causeResponseStripped = exports.causeResponse = exports.TypeId = exports.ServeError = exports.RouteNotFound = exports.ResponseError = exports.RequestError = void 0;
var _Error = require("./Error.js");
var Respondable = _interopRequireWildcard(require("./HttpServerRespondable.js"));
var ServerResponse = _interopRequireWildcard(require("./HttpServerResponse.js"));
var internal = _interopRequireWildcard(require("./internal/httpServerError.js"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 * @category type id
 */
const TypeId = exports.TypeId = internal.TypeId;
/**
 * @since 1.0.0
 * @category error
 */
class RequestError extends /*#__PURE__*/(0, _Error.TypeIdError)(TypeId, "RequestError") {
  /**
   * @since 1.0.0
   */
  [Respondable.symbol]() {
    return ServerResponse.empty({
      status: 400
    });
  }
  get methodAndUrl() {
    return `${this.request.method} ${this.request.url}`;
  }
  get message() {
    return this.description ? `${this.reason}: ${this.description} (${this.methodAndUrl})` : `${this.reason} error (${this.methodAndUrl})`;
  }
}
/**
 * @since 1.0.0
 * @category predicates
 */
exports.RequestError = RequestError;
const isServerError = exports.isServerError = internal.isServerError;
/**
 * @since 1.0.0
 * @category error
 */
class RouteNotFound extends /*#__PURE__*/(0, _Error.TypeIdError)(TypeId, "RouteNotFound") {
  constructor(options) {
    super(options);
    this.stack = `${this.name}: ${this.message}`;
  }
  /**
   * @since 1.0.0
   */
  [Respondable.symbol]() {
    return ServerResponse.empty({
      status: 404
    });
  }
  get message() {
    return `${this.request.method} ${this.request.url} not found`;
  }
}
/**
 * @since 1.0.0
 * @category error
 */
exports.RouteNotFound = RouteNotFound;
class ResponseError extends /*#__PURE__*/(0, _Error.TypeIdError)(TypeId, "ResponseError") {
  /**
   * @since 1.0.0
   */
  [Respondable.symbol]() {
    return ServerResponse.empty({
      status: 500
    });
  }
  get methodAndUrl() {
    return `${this.request.method} ${this.request.url}`;
  }
  get message() {
    const info = `${this.response.status} ${this.methodAndUrl}`;
    return this.description ? `${this.description} (${info})` : `${this.reason} error (${info})`;
  }
}
/**
 * @since 1.0.0
 * @category error
 */
exports.ResponseError = ResponseError;
class ServeError extends /*#__PURE__*/(0, _Error.TypeIdError)(TypeId, "ServeError") {}
/**
 * @since 1.0.0
 */
exports.ServeError = ServeError;
const clientAbortFiberId = exports.clientAbortFiberId = internal.clientAbortFiberId;
/**
 * @since 1.0.0
 */
const causeResponse = exports.causeResponse = internal.causeResponse;
/**
 * @since 1.0.0
 */
const causeResponseStripped = exports.causeResponseStripped = internal.causeResponseStripped;
/**
 * @since 1.0.0
 */
const exitResponse = exports.exitResponse = internal.exitResponse;
